GoogleDataTransport/GDTCORLibrary/GDTCORDatabase.m

```diff
@@ -0,0 +1,316 @@
+  if (cacheStmt) {
+    GDTCORSQLReset(stmt);
+  } else {
+    GDTCORSQLFinalize(stmt);
+  }
+  return result;
+}
+
+/** Runs a series of migrations, setting the user_version PRAGMA to the highest version number.
+ *
+ * @param db The db to operate on.
+ * @param migrations The map of user_version numbers to the SQL statement needed to update to that
+ *   version.
+ * @param stmtCache The statement cache.
+ */
+static void RunMigrations(sqlite3 *db,
+                          NSDictionary<NSNumber *, NSString *> *migrations,
+                          CFMutableDictionaryRef stmtCache) {
+  if (!db) {
+    GDTCORLogError(GDTCORMCEDatabaseError, @"%@", @"The database is closed");
+    return;
+  }
+  int userVersion = GetUserVersion(db);
+  __block int newUserVersion = userVersion;
+  [migrations enumerateKeysAndObjectsUsingBlock:^(NSNumber *_Nonnull key, NSString *_Nonnull obj,
+                                                  BOOL *_Nonnull stop) {
+    if (key.intValue > newUserVersion) {
+      if (RunNonQuery(db, nil, obj, stmtCache, NO)) {
+        newUserVersion = key.intValue;
+      } else {
+        GDTCORLogError(GDTCORMCEDatabaseError, @"%@", @"Migration failed: version:%@ statement:%@",
+                       key, obj);
+      }
+    }
+  }];
+
+  if (newUserVersion != userVersion) {
+    SetUserVersion(db, newUserVersion);
+  }
+}
+
+#pragma mark - GDTCORDatabase
+
+@implementation GDTCORDatabase {
+  /** The sqlite database. */
+  sqlite3 *_db;
+}
+
+- (nullable instancetype)initWithURL:(nullable NSURL *)dbFileURL
+                 migrationStatements:
+                     (nonnull NSDictionary<NSNumber *, NSString *> *)migrationStatements {
+  // Only 1 instance per file is allowed. Return nil if one is already open.
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    instancesToDBFiles = [[NSMutableDictionary alloc] init];
+  });
+  __block GDTCORDatabase *extantDB;
+  dispatch_barrier_sync(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
+    extantDB = instancesToDBFiles[dbFileURL.path];
+  });
+  if (extantDB) {
+    return nil;
+  }
+
+  // Create an instance.
+  self = [super init];
+  if (self) {
+    NSString *dbPath = dbFileURL ? dbFileURL.path : @":memory:";
+    if (!GDTCORSQLOpenDB(&_db, dbPath)) {
+      return nil;
+    }
+    _dbQueue = dispatch_queue_create("com.google.GDTCORDatabase", DISPATCH_QUEUE_SERIAL);
+    _fileURL = dbFileURL;
+    _stmtCache =
+        CFDictionaryCreateMutable(CFAllocatorGetDefault(), 0, &kCFTypeDictionaryKeyCallBacks, NULL);
+    dispatch_async(_dbQueue, ^{
+      RunMigrations(self->_db, [migrationStatements copy], self -> _stmtCache);
+    });
+    dispatch_barrier_sync(dispatch_get_global_queue(QOS_CLASS_DEFAULT, 0), ^{
+      instancesToDBFiles[dbPath] = self;
+    });
+  }
+  return self;
+}
+
+- (void)dealloc {
+  [self close];
+}
+
+- (BOOL)open {
+  if (_db == NULL) {
+    return GDTCORSQLOpenDB(&_db, _fileURL ? _fileURL.path : @":memory:");
```

Comment:
Should it be run on `_dbQueue ` as well?

Reply:
Good catch.